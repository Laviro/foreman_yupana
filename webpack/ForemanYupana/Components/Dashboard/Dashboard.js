import React from 'react';
import PropTypes from 'prop-types';
import { noop } from 'patternfly-react';
import ReportGenerate from '../ReportGenerate';
import ReportUpload from '../ReportUpload';
import NavContainer from '../NavContainer';
import './dashboard.scss';

class Dashboard extends React.Component {
  componentDidMount() {
    const { startPolling, fetchLogs } = this.props;
    const pollingProcessID = setInterval(fetchLogs, 2000);
    startPolling(pollingProcessID);
  }

  componentWillUnmount() {
    const { stopPolling } = this.props;
    stopPolling();
  }

  render() {
    const { setActiveTab } = this.props;
    return (
      <NavContainer
        items={[
          {
            icon: 'database',
            name: 'Generating',
            component: ReportGenerate,
          },
          {
            icon: 'cloud-upload',
            name: 'Uploading',
            component: ReportUpload,
          },
        ]}
        onTabClick={setActiveTab}
      />
    );
  }
}

Dashboard.propTypes = {
  startPolling: PropTypes.func,
  fetchLogs: PropTypes.func,
  stopPolling: PropTypes.func,
  setActiveTab: PropTypes.func,
};

Dashboard.defaultProps = {
  startPolling: noop,
  fetchLogs: noop,
  stopPolling: noop,
  setActiveTab: noop,
};

export default Dashboard;
