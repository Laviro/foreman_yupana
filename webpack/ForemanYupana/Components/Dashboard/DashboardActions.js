import API from 'foremanReact/API';
import {
  YUPANA_POLLING_START,
  YUPANA_POLLING_STOP,
  YUPANA_POLLING,
  YUPANA_TAB_CHANGED,
  YUPANA_POLLING_ERROR,
  YUPANA_PROCESS_RESTART,
  YUPANA_REPORTS_DOWNLOAD,
} from './DashboardConstants';
import { selectActiveTab } from './DashboardSelectors';

export const startPolling = (accountID, pollingProcessID) => ({
  type: YUPANA_POLLING_START,
  payload: {
    accountID,
    pollingProcessID,
  },
});

export const stopPolling = (accountID, pollingProcessID) => dispatch => {
  clearInterval(pollingProcessID);
  dispatch({
    type: YUPANA_POLLING_STOP,
    payload: {
      accountID,
    },
  });
};

export const fetchLogs = accountID => async (dispatch, getState) => {
  const activeTab = selectActiveTab(getState(), accountID);
  try {
    const processController = activeTab === 'uploading' ? 'uploads' : 'reports';
    const {
      data: { output, status },
    } = await API.get(`${accountID}/${processController}/last`);
    dispatch({
      type: YUPANA_POLLING,
      payload: {
        accountID,
        activeTab,
        logs: output,
        exitCode: status,
      },
    });
  } catch (error) {
    dispatch({
      type: YUPANA_POLLING_ERROR,
      payload: {
        accountID,
        activeTab,
        error: error.message,
      },
    });
  }
};

export const setActiveTab = (accountID, tabName) => ({
  type: YUPANA_TAB_CHANGED,
  payload: {
    accountID,
    activeTab: tabName,
  },
});

export const restartProcess = accountID => (dispatch, getState) => {
  const activeTab = selectActiveTab(getState(), accountID);
  const processController = activeTab === 'uploading' ? 'uploads' : 'reports';
  API.post(`${accountID}/${processController}`);
  dispatch({
    type: YUPANA_PROCESS_RESTART,
    payload: {
      accountID,
    },
  });
};

export const downloadReports = accountID => {
  window.location.href = `/foreman_yupana/${accountID}/uploads/file`;
  return {
    type: YUPANA_REPORTS_DOWNLOAD,
    payload: {
      accountID,
    },
  };
};
