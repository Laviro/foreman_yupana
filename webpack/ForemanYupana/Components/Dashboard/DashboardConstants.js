export const YUPANA_POLLING = 'YUPANA_POLLING';
export const YUPANA_POLLING_START = 'YUPANA_POLLING_START';
export const YUPANA_POLLING_STOP = 'YUPANA_POLLING_STOP';
export const YUPANA_POLLING_ERROR = 'YUPANA_POLLING_ERROR';
export const YUPANA_TAB_CHANGED = 'YUPANA_TAB_CHANGED';
export const YUPANA_REPORTS_DOWNLOAD = 'YUPANA_REPORTS_DOWNLOAD';
export const YUPANA_PROCESS_RESTART = 'YUPANA_PROCESS_RESTART';
export const YUPANA_QUEUE = 'YUPANA_QUEUE';
export const YUPANA_QUEUE_ERROR = 'YUPANA_QUEUE_ERROR';
