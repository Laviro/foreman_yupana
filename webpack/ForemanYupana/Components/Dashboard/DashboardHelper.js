export const seperator = '--------------------';

export const addLogs = (processName, logs) => {
  window.__yupana__[processName].logs.push(...logs);
};
