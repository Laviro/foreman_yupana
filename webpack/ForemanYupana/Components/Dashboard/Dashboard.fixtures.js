import { seperator } from './DashboardHelper';

export const pollingProcessID = 1;

export const logs = ['some-logs...'];

export const completed = 25;

export const yupana = {
  generating: {
    logs,
    completed,
  },
  uploading: {
    logs,
    completed,
    files,
  },
};

export const files = ['some-file'];

export const activeTab = 'uploading';

const initialLog = ['No running process', seperator];

export const initialState = {
  generating: {
    logs: initialLog,
    completed,
  },
  uploading: {
    logs: initialLog,
    completed,
    files,
  },
  pollingProcessID: 0,
  activeTab,
};
