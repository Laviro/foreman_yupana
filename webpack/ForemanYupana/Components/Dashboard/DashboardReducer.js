import Immutable from 'seamless-immutable';

import { YUPANA_POLLING_START, YUPANA_POLLING } from './DashboardConstants';
import { seperator } from './DashboardHelper';

const initialState = Immutable({
  /** insert Dashboard state here */
  logs: {
    generating: ['No running process', seperator],
    uploading: ['No running process', seperator],
  },
  completed: {
    generating: 0,
    uploading: 0,
  },
  pollingProcessID: 0,
  files: [],
});

export default (state = initialState, action) => {
  const { payload: { pollingProcessID, logs, completed, files } = {} } = action;

  switch (action.type) {
    case YUPANA_POLLING_START:
      return state.merge({
        pollingProcessID,
      });
    case YUPANA_POLLING: {
      return state.merge({
        logs,
        completed,
        files,
      });
    }
    default:
      return state;
  }
};
