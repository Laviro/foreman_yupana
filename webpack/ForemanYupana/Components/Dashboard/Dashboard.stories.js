import React from 'react';
import { storiesOf } from '@storybook/react';
import { withKnobs } from '@storybook/addon-knobs';
import { Provider } from 'react-redux';
import configureStore from '../../../stories/configureStore';
import Dashboard from './index';
import { generating, uploading } from './Dashboard.fixtures';

const store = configureStore();

storiesOf('Dashboard', module)
  .addDecorator(withKnobs)
  .add('Dashboard', () => (
    <div style={{ margin: '20px' }}>
      <Provider store={store}>
        <Dashboard generating={generating} uploading={uploading} />
      </Provider>
    </div>
  ));
