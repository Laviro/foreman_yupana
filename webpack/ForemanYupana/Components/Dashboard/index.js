import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';

import * as actions from './DashboardActions';
import reducer from './DashboardReducer';

import Dashboard from './Dashboard';

// map state to props
const mapStateToProps = state => ({});

// map action dispatchers to props
const mapDispatchToProps = dispatch => bindActionCreators(actions, dispatch);

// export reducers
export const reducers = { dashboard: reducer };

// export connected component
export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Dashboard);
