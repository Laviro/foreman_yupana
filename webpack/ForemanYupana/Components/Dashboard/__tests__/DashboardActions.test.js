import { testActionSnapshotWithFixtures } from 'react-redux-test-utils';
import API from 'foremanReact/API';
import {
  startPolling,
  stopPolling,
  fetchLogs,
  setActiveTab,
  downloadReports,
  restartProcess,
} from '../DashboardActions';
import { pollingProcessID, serverMock, activeTab } from '../Dashboard.fixtures';

jest.mock('foremanReact/API');
API.get.mockImplementation(() => serverMock);

const runWithGetState = (state, action, params) => dispatch => {
  const getState = () => ({
    dashboard: state,
  });
  action(params)(dispatch, getState);
};

const fixtures = {
  'should startPolling': () => startPolling(pollingProcessID),
  'should fetchLogs': () =>
    runWithGetState({ activeTab: 'uploads' }, fetchLogs),
  'should stopPolling': () => stopPolling(pollingProcessID),
  'should setActiveTab': () => setActiveTab(activeTab),
  'should downloadReports': () => downloadReports(),
  'should restartProcess': () => restartProcess(),
};

describe('Dashboard actions', () => testActionSnapshotWithFixtures(fixtures));
