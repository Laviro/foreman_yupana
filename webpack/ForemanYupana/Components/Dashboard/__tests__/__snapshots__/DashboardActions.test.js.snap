// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard actions should downloadReports 1`] = `
Object {
  "payload": Object {
    "accountID": "some-account-ID",
  },
  "type": "YUPANA_REPORTS_DOWNLOAD",
}
`;

exports[`Dashboard actions should fetchLogs 1`] = `
Array [
  Array [
    Object {
      "payload": Object {
        "accountID": "some-account-ID",
        "activeTab": undefined,
        "exitCode": 0,
        "logs": Array [
          "some-logs...",
        ],
      },
      "type": "YUPANA_POLLING",
    },
  ],
]
`;

exports[`Dashboard actions should restartProcess 1`] = `
Array [
  Array [
    Object {
      "payload": Object {
        "accountID": "some-account-ID",
      },
      "type": "YUPANA_PROCESS_RESTART",
    },
  ],
]
`;

exports[`Dashboard actions should setActiveTab 1`] = `
Object {
  "payload": Object {
    "accountID": "some-account-ID",
    "activeTab": "uploads",
  },
  "type": "YUPANA_TAB_CHANGED",
}
`;

exports[`Dashboard actions should startPolling 1`] = `
Object {
  "payload": Object {
    "accountID": "some-account-ID",
    "pollingProcessID": 1,
  },
  "type": "YUPANA_POLLING_START",
}
`;

exports[`Dashboard actions should stopPolling 1`] = `
Array [
  Array [
    Object {
      "payload": Object {
        "accountID": "some-account-ID",
      },
      "type": "YUPANA_POLLING_STOP",
    },
  ],
]
`;
